### PR TITLE
server: Add DTO beans for /api-violations request/response (#429)

### PR DESCRIPTION
**Pull request by innokenty**
These changes simplify the work with `JsonNode` objects via replacing them with DTO beans with predefined field names.

### CODE REVIEW

File: server/src/test/java/de/zalando/zally/apireview/RestApiEmptyMessageTest.java [hunk 1]
```diff
@@ -28,7 +28,7 @@ public void shouldNotContainMessageFieldWhenMessageIsEmpty() throws IOException
 
         JsonNode rootObject = responseEntity.getBody();
         assertThat(rootObject.has("violations")).isTrue();
-
-        assertThat(rootObject.has("message")).isFalse();
+        assertThat(rootObject.has("message")).isTrue();
+        assertThat(rootObject.get("message").asText()).isEmpty();
```
Comment (innokenty) [inline server/src/test/java/de/zalando/zally/apireview/RestApiEmptyMessageTest.java]: Question to project owners: is this really a feature that if `"message"` property is set to empty — then it's not included in the response at all? I assumed it doesn't really matter and removed this functionality for better code clarity. Did I do right?

**Comment by netme** [comment]
Hey @innokenty , thanks for your contribution. To be compliant with company standards, we need to:
- Add a ticket to GH Issues
- Mention this ticket in PR title / body and commit message.

Could you please take care of that? I will review the code meanwhile.

File: server/src/main/java/de/zalando/zally/apireview/ApiReview.java [hunk 2]
```diff
@@ -1,20 +1,13 @@
...
 
-import javax.persistence.CascadeType;
-import javax.persistence.Column;
-import javax.persistence.Entity;
-import javax.persistence.FetchType;
-import javax.persistence.GeneratedValue;
-import javax.persistence.GenerationType;
-import javax.persistence.Id;
-import javax.persistence.OneToMany;
+import javax.persistence.*;
```
Comment (netme) [inline server/src/main/java/de/zalando/zally/apireview/ApiReview.java]: Could you please configure your IDE to disable star imports?

**Comment by innokenty** [comment]
Should I rename the commit and force-push or PR title is enough?

**Comment by innokenty** [comment]
@netme created, please check: https://github.com/zalando-incubator/zally/issues/429

File: server/src/test/java/de/zalando/zally/apireview/RestApiEmptyMessageTest.java (hunk 1 above)
Comment (netme) [inline server/src/test/java/de/zalando/zally/apireview/RestApiEmptyMessageTest.java]: @innokenty , that should be fine to make it optional according to our Swagger definition: https://github.com/zalando-incubator/zally/blob/master/server/src/main/resources/api/zally-api.yaml#L194 

Could you please make check that web-ui and CLI will accept this behaviour?

File: server/src/main/java/de/zalando/zally/apireview/ApiReview.java (hunk 2 above)
Comment (innokenty) [inline server/src/main/java/de/zalando/zally/apireview/ApiReview.java]: Sure, if the company style says so. I'm just using IDEA defaults.

File: server/src/main/java/de/zalando/zally/violation/ApiDefinitionResponse.kt [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package de.zalando.zally.violation
+
+/**
+ * @author innokenty
+ */
+data class ApiDefinitionResponse (
+
+    var message: String? = null,
+    var violations: List<Violation>? = null,
```
Comment (netme) [inline server/src/main/java/de/zalando/zally/violation/ApiDefinitionResponse.kt]: Violation list is not empty as far as I remember

File: server/src/main/java/de/zalando/zally/violation/ApiDefinitionResponse.kt [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+    var violationsCount: Map<String, Int>? = null
```
Comment (netme) [inline server/src/main/java/de/zalando/zally/violation/ApiDefinitionResponse.kt]: Same for `violationsCount`

File: server/src/main/java/de/zalando/zally/violation/ApiDefinitionRequest.kt [hunk 5]
```diff
@@ -0,0 +1,17 @@
+package de.zalando.zally.violation
+
+import com.fasterxml.jackson.annotation.JsonRawValue
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize
+import de.zalando.zally.util.JsonRawValueDeserializer
+
+/**
+ * @author innokenty
+ */
```
Comment (netme) [inline server/src/main/java/de/zalando/zally/violation/ApiDefinitionRequest.kt]: Normally we are not leaving such comments :-)

File: server/src/main/java/de/zalando/zally/apireview/ApiViolationsController.java [hunk 6]
```diff
@@ -79,10 +65,19 @@ private String retrieveApiDefinition(JsonNode request) {
...
+
+    private Map<String, Integer> buildViolationsCount(List<Violation> violations) {
         ViolationsCounter counter = new ViolationsCounter(violations);
-        for (ViolationType violationType : ViolationType.values()) {
-            result.put(violationType.toString().toLowerCase(), counter.getCounter(violationType));
-        }
+        return Arrays.stream(ViolationType.values()).collect(toMap(
+                violationType -> violationType.toString().toLowerCase(),
+                counter::getCounter
+        ));
```
Comment (netme) [inline server/src/main/java/de/zalando/zally/apireview/ApiViolationsController.java]: That's nice

**Comment by innokenty** [comment]
@netme 
- fixed star imports
- added issue in commit name
- force-pushed with these changes

File: server/src/test/java/de/zalando/zally/apireview/RestApiEmptyMessageTest.java (hunk 1 above)
Comment (innokenty) [inline server/src/test/java/de/zalando/zally/apireview/RestApiEmptyMessageTest.java]: Ough.. I guess I gonna need a tutorial on how to do it. I'll try.

**Comment by netme** [comment]
@innokenty force push would be great

**Comment by innokenty** [comment]
@netme already did, please refresh the page to see my comment above 😉

File: server/src/main/java/de/zalando/zally/violation/ApiDefinitionRequest.kt (hunk 5 above)
Comment (innokenty) [inline server/src/main/java/de/zalando/zally/violation/ApiDefinitionRequest.kt]: Okay, I'll remove. Just my IDE config for new files.

File: server/src/test/java/de/zalando/zally/apireview/RestApiEmptyMessageTest.java (hunk 1 above)
Comment (netme) [inline server/src/test/java/de/zalando/zally/apireview/RestApiEmptyMessageTest.java]: I'm pretty sure that web-ui is not using it, you can check zalando-incubator/zally-web-ui . For CLI please check CLI implementation from `cli/` folder.

**Comment by netme** [comment]
@innokenty you were faster than me ;-)

File: server/src/main/java/de/zalando/zally/violation/ApiDefinitionResponse.kt (hunk 3 above)
Comment (innokenty) [inline server/src/main/java/de/zalando/zally/violation/ApiDefinitionResponse.kt]: Yeah, but it's set as a setter later, not at the constructor stage. Usually DTO beans are allowed have any field with a default java value `null` as well as the empty constructor. For unmarshalling reasons and so on.

Ah wait, you mean in the Zally API doc?

File: server/src/main/java/de/zalando/zally/violation/ApiDefinitionResponse.kt (hunk 4 above)
Comment (innokenty) [inline server/src/main/java/de/zalando/zally/violation/ApiDefinitionResponse.kt]: If I understand you correctly: only the `violations` field is required:

https://github.com/zalando-incubator/zally/blob/master/server/src/main/resources/api/zally-api.yaml#L186

Comment (netme) [inline server/src/main/java/de/zalando/zally/violation/ApiDefinitionResponse.kt]: hm.. actually you are right, only `violations` are required.

File: server/src/main/java/de/zalando/zally/violation/ApiDefinitionResponse.kt (hunk 3 above)
Comment (netme) [inline server/src/main/java/de/zalando/zally/violation/ApiDefinitionResponse.kt]: Yes

File: server/src/main/java/de/zalando/zally/violation/Violation.kt [hunk 7]
```diff
@@ -1,13 +1,12 @@
...
 data class Violation(
     @JsonIgnore val rule: Rule,
     val title: String,
     val description: String,
-    @JsonProperty("violation_type") val violationType: ViolationType,
-    @JsonProperty("rule_link") val ruleLink: String,
-    val paths: List<String>)
\ No newline at end of file
+    val violationType: ViolationType,
+    val ruleLink: String,
```
Comment (netme) [inline server/src/main/java/de/zalando/zally/violation/Violation.kt]: Where the snake_case transformation will happen?

Comment (innokenty) [inline server/src/main/java/de/zalando/zally/violation/Violation.kt]: Here: https://github.com/zalando-incubator/zally/blob/master/server/src/main/java/de/zalando/zally/configuration/JacksonObjectMapperConfiguration.java#L23

I removed these annotations and saw that nothing fails. Also, my newly introduced class doesn't have those annotations either, but it definitely transforms to a snake_cased json.

Comment (netme) [inline server/src/main/java/de/zalando/zally/violation/Violation.kt]: Ah, I've missed that, sorry.

**Comment by netme** [comment]
👍

**Comment by netme** [comment]
👍

**Pull request merged**